src/java/picard/sam/PositionBasedDownsampleSam.java

```diff
@@ -0,0 +1,361 @@
+
+        final double finalP = kept / (double) total;
+        if (Math.abs(finalP - PROBABILITY) / (Math.min(finalP, PROBABILITY) + 1e-10) > .2) {
+            log.warn(String.format("You've requested PROBABILITY=%g, the resulting downsampling resulted in a rate of %f.", PROBABILITY, finalP));
+        }
+        log.info(String.format("Finished! Kept %d out of %d reads (P=%g).", kept, total, finalP));
+
+        return 0;
+    }
+
+    private void outputSamRecords() {
+
+        final ProgressLogger progress = new ProgressLogger(log, (int) 1e7);
+        final SamReader in = SamReaderFactory.makeDefault().referenceSequence(REFERENCE_SEQUENCE).open(INPUT);
+
+        final SAMFileHeader header = in.getFileHeader().clone();
+        final SAMFileHeader.PgIdGenerator pgIdGenerator = new SAMFileHeader.PgIdGenerator(header);
+        final SAMProgramRecord pr = new SAMProgramRecord(pgIdGenerator.getNonCollidingId(PG_PROGRAM_NAME));
+
+        pr.setProgramName(PG_PROGRAM_NAME);
+        pr.setCommandLine(getCommandLine());
+        pr.setProgramVersion(getVersion());
+        header.addProgramRecord(pr);
+
+        final SAMFileWriter out = new SAMFileWriterFactory().makeSAMOrBAMWriter(header, true, OUTPUT);
+
+        final CircleSelector selector = new CircleSelector(PROBABILITY);
+
+        for (final SAMRecord rec : in) {
+            if (STOP_AFTER != 0 && total >= STOP_AFTER) break;
+
+            total++;
+
+            final PhysicalLocation pos = getSamRecordLocation(rec);
+
+            if (!xPositions.containsKey(pos.getTile())) {
+                xPositions.put(pos.getTile(), new Histogram<Short>(pos.getTile() + "-xpos", "count"));
+            }
+            if (!yPositions.containsKey(pos.getTile())) {
+                yPositions.put(pos.getTile(), new Histogram<Short>(pos.getTile() + "-ypos", "count"));
+            }
+
+            final boolean keepRecord = selector.select(pos, tileCoord.get(pos.getTile()));
+
+            if (keepRecord) {
+                if (REMOVE_DUPLICATE_INFORMATION) rec.setDuplicateReadFlag(false);
+                out.addAlignment(rec);
+                kept++;
+            }
+            progress.record(rec);
+        }
+
+        out.close();
+
+        CloserUtil.close(in);
+    }
+
+    private void checkProgramRecords() {
+        final SamReader in = SamReaderFactory
+                .makeDefault()
+                .referenceSequence(REFERENCE_SEQUENCE)
+                .open(INPUT);
+
+        for (final SAMProgramRecord pg : in.getFileHeader().getProgramRecords()) {
+            if (pg.getProgramName() != null && pg.getProgramName().equals(PG_PROGRAM_NAME)) {
+
+                String outText = "Found previous Program Record that indicates that this BAM has been downsampled already with this program. Operation not supported! Previous PG: " + pg.toString();
+
+                if (ALLOW_MULTIPLE_DOWNSAMPLING_DESPITE_WARNINGS) {
+                    log.warn(outText);
```

Comment:
Despite the warning that will say downsampling again is impossible, isn't this going to downsample again if `ALLOW...DESPITE_WARNINGS=true` is specified?